### ISSUE TITLE
make it allocation free on happy path

### ISSUE THREAD

**Issue by nevgeniev** [wrandelshofer/FastDoubleParser#11]
https://github.com/wrandelshofer/FastDoubleParser/blob/8df68712f159304a828b4ddfc11a14a54ec6c973/src/main/java/ch/randelshofer/fastdoubleparser/FastDoubleParser.java#L530

why allocate here?? by that time you know it's not a NaN for sure... so instead of returning `null` you can just return Double.NaN or whatever special constant.

lack of jmh tests is also troubling :(

**Comment by lemire** [comment]
(Note that I am not one of the authors of this library, just an observer.)

> why allocate here?? 

Are you sure that there is heap allocation? What does [escape analysis reveals](https://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#escapeAnalysis)?


> lack of jmh tests is also troubling :(


Presumably, you are unsatisfied by `DoubleParserJmhBenchmark.java`?

**Comment by wrandelshofer** [comment]
Hi nevgeniev,

I did indeed implement it first with a NaN that used some of the available bits in the NaN to encode that we need to go over the slow path. However this did not improve performance. It just made the code more complex.
Heap allocation occurs actually only in interpreted mode. The JIT inlines the called method and then removes the allocation. -In the use cases, in which I use FastDoubleParser, there is no point in optimizing interpreted mode.

Actually, there are jmh benchmarks in the test/jmh/java folder.

**Comment by nevgeniev** [comment]
@lemire ea could be very fragile actually... for it to work function allocating Double needs to be inlined first. And for this to happen certain hotspot threshold must not be hit. (max number of  instructions already inlined and stack depth) if those threshold were hit, hotspot will just stop inlining, so function allocating Double will become just a function call and actuall allocation would happen at least this is my understanding :). For example because of that effect Unsafe in latest (8+) jdk versions works slower for some libraries (stacks just became deeper as `sun.misc.Unsafe` became a facade rather than real Unsafe).. and to win performance back you need to adjust those thresholds

tl;dr, personally I prefer to make code explicitly alloc free whenever is possible to avoid surprises

**Comment by nevgeniev** [comment]
@wrandelshofer yeah sorry I didn't notice, output in README was not from jmh this is why :(

**Comment by lemire** [comment]
@nevgeniev By very fragile... you mean that it is easy to cause heap allocation, even in the case where the function is called enough to be inlined.

This is very interesting. Can you provide me with such an example (an actual benchmark)... I'd love to take a look.

**Comment by wrandelshofer** [comment]
I have made the code now allocation free on the happy path in revision 343e4aeb862eb792e6d0c10c616088ea97752289.
